File: src/main/java/com/DavidM1A2/afraidofthedark/common/spell/component/effect/base/AOTDSpellEffect.java
package com.DavidM1A2.afraidofthedark.common.spell.component.effect.base;

import com.DavidM1A2.afraidofthedark.AfraidOfTheDark;
import com.DavidM1A2.afraidofthedark.client.particle.AOTDParticleRegistry;
import com.DavidM1A2.afraidofthedark.common.packets.otherPackets.SyncParticle;
import net.minecraft.util.math.Vec3d;
import net.minecraftforge.fml.common.network.NetworkRegistry;
import org.apache.commons.lang3.RandomUtils;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

/**
 * Base class for all AOTD effects
 */
public abstract class AOTDSpellEffect extends SpellEffect
{
    /**
     * Constructor just calls super currently
     */
    public AOTDSpellEffect()
    {
        super();
    }

    /**
     * Creates particles at the position. This is static so overridden effects can still use it
     *
     * @param min       The minimum number of particles to spawn
     * @param max       The maximum number of particles to spawn
     * @param pos       The position to spawn particles at
     * @param dimension The dimension to create particles in
     */
    public static void createParticlesAt(int min, int max, Vec3d pos, int dimension)
    {
        // Spawn particles
        List<Vec3d> positions = new ArrayList<>();
        for (int i = 0; i < RandomUtils.nextInt(min, max + 1); i++)
        {
            positions.add(pos);
        }
        // Send the particle packet
        AfraidOfTheDark.INSTANCE.getPacketHandler().sendToAllAround(
                new SyncParticle(AOTDParticleRegistry.ParticleTypes.SPELL_HIT_ID, positions, Collections.nCopies(positions.size(), Vec3d.ZERO)),
                new NetworkRegistry.TargetPoint(dimension, pos.x, pos.y, pos.z, 100));
    }
}

File: src/main/java/com/davidm1a2/afraidofthedark/common/spell/component/effect/base/AOTDSpellEffect.kt
package com.davidm1a2.afraidofthedark.common.spell.component.effect.base

import com.davidm1a2.afraidofthedark.AfraidOfTheDark
import com.davidm1a2.afraidofthedark.client.particle.AOTDParticleRegistry
import com.davidm1a2.afraidofthedark.common.packets.otherPackets.SyncParticle
import net.minecraft.util.ResourceLocation
import net.minecraft.util.math.Vec3d
import net.minecraftforge.fml.common.network.NetworkRegistry.TargetPoint
import kotlin.random.Random

/**
 * Base class for all AOTD effects
 *
 * @param id The id of the spell effect
 * @constructor just calls super currently
 */
abstract class AOTDSpellEffect(id: ResourceLocation) : SpellEffect(id)
{
    companion object
    {
        /**
         * Creates particles at the position. This is static so overridden effects can still use it
         *
         * @param min       The minimum number of particles to spawn
         * @param max       The maximum number of particles to spawn
         * @param pos       The position to spawn particles at
         * @param dimension The dimension to create particles in
         */
        fun createParticlesAt(min: Int, max: Int, pos: Vec3d, dimension: Int)
        {
            // Spawn particles
            val positions = List(Random.nextInt(min, max + 1)) { pos }

            // Send the particle packet
            AfraidOfTheDark.INSTANCE.packetHandler.sendToAllAround(
                    SyncParticle(AOTDParticleRegistry.ParticleTypes.SPELL_HIT_ID, positions, List(positions.size) { Vec3d.ZERO }),
                    TargetPoint(dimension, pos.x, pos.y, pos.z, 100.0)
            )
        }
    }
}